File: Graph/Python/Find Minimum Diameter After Merging Two Trees.py
# Problem link: https://leetcode.com/problems/find-minimum-diameter-after-merging-two-trees/description
# Time: O(n + m)
# Space: O(n + m)
class Solution:
    def minimumDiameterAfterMerge(self, edges1: List[List[int]], edges2: List[List[int]]) -> int:
        adjListOne = [[] for _ in range(len(edges1) + 1)]
        adjListTwo = [[] for _ in range(len(edges2) + 1)]

        for a, b in edges1:
            adjListOne[a].append(b)
            adjListOne[b].append(a)
        
        for a, b in edges2:
            adjListTwo[a].append(b)
            adjListTwo[b].append(a)
        
        diameterOne = self.maxDiameter(-1, 0, adjListOne)[0]
        diameterTwo = self.maxDiameter(-1, 0, adjListTwo)[0]

        return max(diameterOne, diameterTwo, 1 + ceil(diameterOne / 2) + ceil(diameterTwo / 2))
    
    def maxDiameter(self, par, node, adjList):
        first = second = maxDia = 0

        for nei in adjList[node]:
            if nei != par:
                neiDia, longestPath = self.maxDiameter(node, nei, adjList)
                maxDia = max(maxDia, neiDia)
                if longestPath > first:
                    second = first
                    first = longestPath
                elif longestPath > second:
                    second = longestPath

        return [max(maxDia, first + second), first + 1]

File: Graph/Java/Find Minimum Diameter After Merging Two Trees.java
// Problem link: https://leetcode.com/problems/find-minimum-diameter-after-merging-two-trees/description/
// Time: O(n + m)
// Space: O(n + m)
class Solution {
    public int minimumDiameterAfterMerge(int[][] edges1, int[][] edges2) {
        List<Integer>[] adjListOne = this.buildAdjList(edges1);
        List<Integer>[] adjListTwo = this.buildAdjList(edges2);

        int diameterOne = this.maxDiameter(-1, 0, adjListOne)[0];
        int diameterTwo = this.maxDiameter(-1, 0, adjListTwo)[0];
        int addGraph = (int)Math.ceil(diameterOne / 2.0) + (int)Math.ceil(diameterTwo / 2.0) + 1;

        return Math.max(addGraph, Math.max(diameterOne, diameterTwo));
    }

    private List<Integer>[] buildAdjList(int[][] edges) {
        int size = edges.length + 1;
        List<Integer>[] adjList = new ArrayList[size];

        for(int index = 0; index < size; index++) {
            adjList[index] = new ArrayList<>();
        }

        for(int[] edge : edges) {
            adjList[edge[0]].add(edge[1]);
            adjList[edge[1]].add(edge[0]);
        }

        return adjList;
    }

    private int[] maxDiameter(int parent, int node, List<Integer>[] adjList) {
        int first = 0, second = 0, maxDia = 0, store[];

        for(int nei : adjList[node]) {
            if(nei != parent) {
                store = this.maxDiameter(node, nei, adjList);
                maxDia = Math.max(maxDia, store[0]);

                if(store[1] > first) {
                    second = first;
                    first = store[1];
                }
                else if(store[1] > second) {
                    second = store[1];
                }
            }
        }

        maxDia = Math.max(maxDia, first + second);
        return new int[] {maxDia, first + 1};
    }
}
